gobblin-utility/src/main/java/org/apache/gobblin/util/hadoop/TokenUtils.java

```diff
@@ -82,6 +88,48 @@
   private static final Pattern KEYTAB_USER_PATTERN = Pattern.compile(".*\\/.*@.*");
   private static final String KERBEROS_REALM = "kerberos.realm";
 
+  /**
+   * the key that will be used to set proper signature for each of the hcat token when multiple hcat
+   * tokens are required to be fetched.
+   */
+  private static final String HIVE_TOKEN_SIGNATURE_KEY = "hive.metastore.token.signature";
+  /**
+   * User can specify the hcat location that they used specifically. It could contains addtional hcat location,
+   * comma-separated.
+   */
+  private static final String USER_DEFINED_HIVE_LOCATIONS = "user.defined.hcatLocation";
+
+  /**
+   * Get Hadoop tokens (tokens for job history server, job tracker, hive and HDFS) using Kerberos keytab,
+   * on behalf on a proxy user, and embed tokens into a {@link UserGroupInformation} as returned result.
+   *
+   * Note that when a super-user is fetching tokens for other users,
+   * {@link #fetchHcatToken(String, HiveConf, String, IMetaStoreClient)} getDelegationToken} explicitly
+   * contains a string parameter indicating proxy user, while other hadoop services requires impersonation first.
+   *
+   * @param state A {@link State} object that should contain properties.
+   * @param tokenFile The file that will finally store materialized credentials.
+   * @param ugi The {@link UserGroupInformation} that to be added with negotiated credentials.
+   * @param targetUser The user to be impersonated as for fetching hadoop tokens.
+   * @return A {@link UserGroupInformation} containing negotiated credentials.
+   */
+  public static UserGroupInformation getHadoopAndHiveTokens(final State state, File tokenFile, UserGroupInformation ugi,
```

Comment:
This seems to be specifically for proxying. What about renaming to `getHadoopAndHiveTokensForProxyUser`